### PR TITLE
[MemCpyOpt] Use EarliestEscapeInfo

### PR DESCRIPTION
**Pull request by nikic**
Pass EarliestEscapeInfo to BatchAA in MemCpyOpt. This allows memcpy elimination in cases where one of the involved pointers is captured after the relevant memcpy/call.

### CODE REVIEW

**Review by dianqk — APPROVED** [review]
LGTM, thanks!
This context is https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/nocapture.20and.20allocation.20elimination.

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp [hunk 1]
```diff
@@ -284,6 +284,7 @@ static bool mayBeVisibleThroughUnwinding(Value *V, Instruction *Start,
 
 void MemCpyOptPass::eraseInstruction(Instruction *I) {
   MSSAU->removeMemoryAccess(I);
+  EEI->removeInstruction(I);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:287]: Should we call `MemCpyOptPass::eraseInstruction` in https://github.com/llvm/llvm-project/blob/a94060ca0c87d13af5c04410f3ed24c01d1de909/llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp#L1157? `NewCopySource` may be deleted after a BAA query.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:287]: Nice catch, fixed!

**Pull request merged**